### PR TITLE
[FFM-10041] - Fall back to identifier if bucketBy attr is not found

### PR DESCRIPTION
**Pull request by andybharness**
[FFM-10041] - Fall back to identifier if bucketBy attribute is not found

**What**
Update BucketBy logic to fall back to the target identifier if given BucketBy attribute is not found.
Adds new unit test to test fallback logic works.

**Why**
Keep the SDK consistent with the behaviour of the other server SDKs.

**Testing**
Testgrid + new unit test

[FFM-10041]: https://harness.atlassian.net/browse/FFM-10041?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by erdirowlands — APPROVED** [review]

File: src/__tests__/evaluator.test.ts
Comment (knagurski) [inline src/__tests__/evaluator.test.ts:1]: Awesome job making sure tests are included!

File: src/evaluator.ts [hunk 1]
```diff
@@ -76,9 +80,19 @@ export class Evaluator {
     bucketBy: string,
     percentage: number,
   ): boolean {
-    const property = this.getAttrValue(target, bucketBy);
+    let property = this.getAttrValue(target, bucketBy);
     if (!property) {
-      return false;
+      const old_bucketBy = bucketBy;
```
Comment (knagurski) [inline src/evaluator.ts:85]: The standard in JS is camel-case, not snake
```suggestion
      const oldBucketBy = bucketBy;
```

File: src/evaluator.ts [hunk 2]
```diff
@@ -76,9 +80,19 @@ export class Evaluator {
...
+      bucketBy = 'identifier';
```
Comment (knagurski) [inline src/evaluator.ts:86]: There's technically no reason you can't do this as args are considered to be `var` (as opposed to `let` or `const`), but it's generally a bad idea to mutate args. Maybe better to extract a `let` variable as you did with `property`

File: src/evaluator.ts [hunk 3]
```diff
@@ -76,9 +80,19 @@ export class Evaluator {
...
+      property = this.getAttrValue(target, bucketBy);
+      if (!property) {
+        return false;
+      }
+      warnBucketByAttributeNotFound(
```
Comment (knagurski) [inline src/evaluator.ts:91]: This is becoming a bit of a wall of code. A few line breaks around the `if` will make it a lot easier to visually scan.
```suggestion
      property = this.getAttrValue(target, bucketBy);
      
      if (!property) {
        return false;
      }
      
      warnBucketByAttributeNotFound(
```

File: src/evaluator.ts [hunk 4]
```diff
@@ -76,9 +80,19 @@ export class Evaluator {
...
+        old_bucketBy,
+        property?.toString(),
+        this.log,
+      );
     }
     const bucketId = this.getNormalizedNumber(property, bucketBy);
```
Comment (knagurski) [inline src/evaluator.ts:97]: If I remember correctly, Node has the property and bucket by backwards in this, can we verify and fix that at the same time?

File: src/__tests__/evaluator.test.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import { Evaluator } from '../evaluator';
+import { Logger } from '../log';
+
+describe('Evaluator', () => {
+  it('should return identifier if bucketby attribute does not exist', async () => {
```
Comment (knagurski) [inline src/__tests__/evaluator.test.ts:5]: Technically this does not return identifier if bucketBy attribute doesn't exist, buckets by the identifier if the bucketBy attribute doesn't exist and returns the appropriate variation value.

File: src/evaluator.ts (hunk 1 above)
Comment (andybharness) [reply src/evaluator.ts:85]: lol - I work on so many languages these days I've no idea anymore

File: src/evaluator.ts (hunk 4 above)
Comment (andybharness) [reply src/evaluator.ts:99]: do you mean to keep parameter ordering consistent consistent with other SDKs? yeah should be easy enough

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@harnessio/ff-nodejs-server-sdk",
-  "version": "1.3.7",
+  "version": "1.3.8",
```
[the remark this answers is not in the document]
Comment (andybharness) [reply package.json:3]: Makes sense - I've been bumping the second digit for other SDKs but only if the hash output changed, NodeJS had the correct hash but needed this fallback code. Should be easy to update

**Review by knagurski — APPROVED** [review]
👍🏻

**Pull request merged**